Split Vulkan from generic bindings, and add DirectX bindings

TimethyH opened the pull request:
Depends on #36

This PR splits the bindings and their implementations into their respective API modules. NGX is used as the common/generic bindings.
Both API's have been tested and function as expected. 
This PR only handles SuperSampling. RayReconstruction is not yet supported on DirectX.


The user now has to call the API specific functions such as vk::System::new() or dx::feature::Feature::new_super_sampling().

Finally this PR updates the DLSS submodule to the latest libs.

MarijnS95 committed:
Link DLSS libraries on Windows

Co-authored-by:  TimethyH <timethy.hyman@hotmail.com>

TimethyH committed:
Separate Directx, Vulkan and Generic bindings.

TimethyH committed:
Separate Directx, Vulkan and Generic bindings.

TimethyH committed:
Cleanup code before PR

TimethyH committed:
Cleanup code before PR

TimethyH committed:
CI Windows Fix and inf recursion

TimethyH committed:
CI Windows Fix and inf recursion

TimethyH committed:
CI Windows Fix

TimethyH committed:
CI Windows Fix

TimethyH committed:
CI fixes. Documented comments and Windows crate resolve issue

TimethyH committed:
CI fixes. Documented comments and Windows crate resolve issue

TimethyH committed:
CI Fix: unsafe function

TimethyH committed:
CI Fix: unsafe function

TimethyH committed:
CI Fix: unsafe function

TimethyH committed:
CI Fix: unsafe function

TimethyH committed:
Ran cargo fmt

TimethyH committed:
Ran cargo fmt

TimethyH committed:
Updatess CI to build using build rather than -b

TimethyH committed:
Updatess CI to build using build rather than -b

TimethyH committed:
CI Fix: Cleans the correct binding directories

TimethyH committed:
CI Fix: Cleans the correct binding directories

TimethyH committed:
CI Fix: Documented lines on commented code

TimethyH committed:
CI Fix: Documented lines on commented code

TimethyH committed:
removes ngx bindings

TimethyH committed:
adds the NGX bindings agianD

TimethyH committed:
Adds clang debug to CI

TimethyH committed:
CI: Build with consistent enum types

MarijnS95 reviewed (commented):
Looks like I had an unfinished review sitting from last week :)

TimethyH committed:
CI: Build with prepend enum name

TimethyH committed:
CI: Build with signed char

TimethyH committed:
CI: Builds with new type default enum

TimethyH committed:
CI VK only test

TimethyH committed:
CI VK only test

TimethyH committed:
Resolves PR feedback from Mariijn

TimethyH committed:
Resolves PR feedback from Mariijn

TimethyH committed:
Merge branch 'api-separation' of https://github.com/Traverse-Research/nvngx-rs into api-separation

TimethyH committed:
Rewrites unsupported platform handling. Gives the user a warning that DLSS is not supported rather than panic

TimethyH committed:
CI Uploads generated bindings

TimethyH committed:
Undid enum change from last commit

TimethyH committed:
CI fixes wrong indentation

TimethyH committed:
CI binding generation uploads and commits the generated bindins.

TimethyH committed:
CI fixes wrong indentation

TimethyH committed:
Uploads the correct bindings

TimethyH committed:
Adds pub parameter destroy fn to cleanup the vk capability params

TimethyH committed:
Removes unused include

TimethyH commented on crates/nvngx/Cargo.toml:
```diff
@@ -12,8 +12,16 @@ repository = "https://github.com/iddm/nvngx-rs"
 [dependencies]
 widestring = "1"
-ash = "0.38"
+ash = { version = "0.38", optional = true }
```
Will leave this for a follow-up PR since i cant get it to work. It gives me the error that Debug is not implemented eventhough i call #[derive(Debug) on the relevant types.

TimethyH committed:
Removes some unused code and specifies that RR needs to be implemented for DX12

TimethyH committed:
runs fmt

TimethyH committed:
Compiles VK header as cpp to prevent syntax error

TimethyH committed:
Supports C++ compilation by wrapping the inlined functions in extern C

TimethyH committed:
Deletes commented lines

TimethyH committed:
Adds ifdef cpp guards for extern C

TimethyH committed:
Checks if the vulkan sdk supports DLSS. If not kindly asks to upgrade to version 1.4 or higher

TimethyH committed:
Runs fmt

TimethyH committed:
Removes unused format macro

TimethyH committed:
Parses the installed vulkan sdk and checks if the dlss requirements are met

TimethyH committed:
Parses SDK version when creating the ngx system. If version is not supported by dlss returns error

TimethyH committed:
Removes SDK version checking. The issue was the validation layer, but if ignored dlss works, meaning we should not force the newer version here

iddm reviewed (commented):
Thank you very much for working on it! I have a few comments. Have a great day!

TimethyH commented on crates/nvngx-sys/src/lib.rs:
```diff
@@ -5,16 +5,15 @@
 pub use error::*;
+
+pub mod vulkan;
+
```
There is a cfg guard within the vulkan.rs and directx.rs files themselves.

TimethyH committed:
Implements resolved PR comments left by iddm

TimethyH committed:
Comment was missplaced, moved it above its respective code

iddm commented on crates/nvngx-sys/src/vk_wrapper.h:
```diff
@@ -0,0 +1,46 @@
+#ifdef __cplusplus
+extern "C"
+{
+    #endif
```
This `#endif` decided to abandon his brother. :)

iddm reviewed (commented):
Hey guys, are you still on it? :-)

iddm commented on crates/nvngx/src/vk/feature.rs:
```diff
@@ -246,222 +89,29 @@ impl FeatureParameters {
     /// with NGX capabilities and available features.
-    pub fn get_capability_parameters() -> Result<Self> {
+    pub fn vk_get_capability_parameters() -> Result<Self> {
         let mut ptr: *mut nvngx_sys::NVSDK_NGX_Parameter = std::ptr::null_mut();
         Result::from(unsafe {
-            nvngx_sys::NVSDK_NGX_VULKAN_GetCapabilityParameters(&mut ptr as *mut _)
+            nvngx_sys::vulkan::NVSDK_NGX_VULKAN_GetCapabilityParameters(&mut ptr as *mut _)
         })
         .map(|_| Self(ptr))
     }
```
Sure thing, good luck!

TimethyH committed:
Start of Trait usage

Not done but pushing so i cann contnue at home

TimethyH committed:
Trait prototype finished

needs to be refined, cleaned up and dx features need to be added but traits should work for vk

TimethyH committed:
runs fmt

TimethyH committed:
CI fix, build docs.

TimethyH committed:
CI fix, documented code

TimethyH committed:
exposes public types

TimethyH commented on crates/nvngx-sys/build.rs:
```diff
@@ -28,40 +122,71 @@ fn vulkan_sdk_include_directory() -> Option<PathBuf> {
         "windows" => "DLSS/lib/Windows_x86_64",
         "linux" => "DLSS/lib/Linux_x86_64",
-        x => todo!("No libraries for {x}"),
+        _ => "cargo:warning=DLSS is not supported on this platform ({}). Skipping link step.",
```
reverted this

TimethyH committed:
Code cleanup following Marijn's review

TimethyH committed:
runs fmt

iddm commented:
Hey guys, please excuse me for my long absence. Is there anything else you want to push to this branch, or I can already take a look and merge it all good?

iddm commented:
I have just taken a quick look, found that the changes on the branch weren't ready yet. Feel free to update. I, perhaps, might take the changes from your branch myself at a later time (especially if you don't respond). Thank you for your help!
